keps/sig-storage/1698-generic-inline-volumes/README.md

```diff
@@ -0,0 +1,561 @@
+
+### Modifying volumes
+
+Once the PVC for an inline volume has been created, it can be updated
+directly like other PVCs. The controller will not interfere with that
+because it never updates PVCs. This can be used to control features
+like volume resizing.
+
+### Late binding
+
+Ideally, provisioning should use late binding (aka
+`WaitForFirstConsumer`). The initial implementation assumes that the
+user is taking care of that by selecting that provisioning mode in the
+storage class for the PVC.
+
+Later, `kube-scheduler` and `external-provisioner` can be changed to
+automatically enable late binding for PVCs which are owned by a pod.
+
+### Test Plan
+
+- Unit tests will be added for the API change.
+- Unit tests will cover the functionality of the controller, similar
+  to
+  https://github.com/kubernetes/kubernetes/blob/v1.18.2/pkg/controller/volume/persistentvolume/pv_controller_test.go.
+- Unit tests need to cover the positive case (feature enabled) as
+  well as negative case (feature disabled or feature used incorrectly).
+- A new [storage test
+  suite](https://github.com/kubernetes/kubernetes/blob/2b2cf8df303affd916bbeda8c2184b023f6ee53c/test/e2e/storage/testsuites/base.go#L84-L94)
+  will be added which tests inline volume creation, usage and deletion
+  in combination with all drivers that support dynamic volume
+  provisioning.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Volume name collisions emitted as pod events
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- 3 examples of real world usage
+- Downgrade tests and scalability tests
+- Allowing time for feedback
+
+### Upgrade / Downgrade Strategy
+
+When downgrading to a cluster which does not support generic inline
+volumes (either by disabling the feature flag or an actual version
+downgrade), pods using generic inline volumes will no longer be
+scheduled or started.
+
+### Version Skew Strategy
+
+As with downgrades, having some of the relevant components at an older
+version will prevent pods from starting.
+
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate
+    - Feature gate name: GenericInlineVolumes
+    - Components depending on the feature gate:
+      - kube-apiserver
+      - kube-controller-manager
+      - kubelet
+
+* **Does enabling the feature change any default behavior?**
+  If users are allowed to create pods but not PVCs, then generic inline volumes
+  grants them permission to create PVCs indirectly. Cluster admins must take
+  that into account in their permission model.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  Yes, by disabling the feature gates. Existing pods will continue to run and
+  volumes will be deleted once they stop. Existing pods with generic inline
```

Comment:
> Existing pods will continue to run and  volumes will be deleted once they stop.

If someone just restarts kubelet with the feature gate disabled, it won't be able to unmount InLine volumes from running Pods. I haven't found documentation how to actually flip a feature gate in a running cluster. Can you please add a note that node must be drained before flipping the feature gate on kubelet, just to be on the safe side?